src/main/java/io/nats/client/impl/NatsConnection.java

```diff
@@ -1508,38 +1517,89 @@ public long getMaxPayload() {
     }
 
     public Collection<String> getServers() {
-        NatsServerInfo info = this.serverInfo.get();
-        HashSet<String> check = new HashSet<>();
-        ArrayList<String> servers = new ArrayList<>();
-
-        options.getServers().stream().forEach(x -> {
-            String uri = x.toString();
-            if (!check.contains(uri)) {
-                servers.add(uri);
-                check.add(uri);
-            }
-        });
+        return getServersList();
+    }
+
+    private List<String> getServersList() {
+        List<String> servers = new ArrayList<>();
+        addConfiguredServers(servers);
+        addDiscoveredServers(servers);
+        return servers;
+    }
+
+    private List<String> getReconnectServerList() {
+        if (options.isNoRandomize()) {
+            return getServersList();
+        }
+
+        // prioritize discovered over configured
+        List<String> servers = new ArrayList<>();
+        addDiscoveredServers(servers);
+        shuffleWithoutCurrent(servers);

```

Comment:
Couldn't shuffle be called once after adding discovered and then configured?

Reply:
The thought here was that discovered servers are the most viable and contain the list of configured servers, although possible in a ip form instead of a dns name form. In fact I'm not even sure of the value of adding the configured servers in the list once the discovered are populated.

Reply:
Discovered servers should be only servers that weren't originally configured.  For example, if I have servers A, B, and C configured in the client then added D and E, the client would have A,B,C as configured servers and D and E as a discovered servers.  Randomization should be applied to the whole list (sans current during reconnect), so each server will be weighted equally in probability w/r/t reconnect order.  Someday we'll have classes of servers for regional preference (e.g. try this set first, then if all those fail, try another set), but that's TBD at the moment.

There is a chance for duplicates based on IP vs hostname, but server configuration allows ops to handle that.